zeppelin-web-angular/e2e/reporter.coverage.ts

```diff
@@ -0,0 +1,304 @@
+
+    return `Tested pages: ${tested}/${results.length} (${testedRate}%) (${this.toRate(success, tested).toFixed(
+      2
+    )}%, success: ${success}, failed: ${failed}, skipped: ${skipped})`;
+  }
+
+  getTestCasesResult(results: ResultsType) {
+    const stats = results.reduce(
+      (acc, item) => ({
+        total: acc.total + item[TABLE_COLUMNS.TOTAL],
+        failed: acc.failed + item[TABLE_COLUMNS.FAILED],
+        skipped: acc.skipped + item[TABLE_COLUMNS.SKIPPED],
+        success: acc.success + item[TABLE_COLUMNS.SUCCESS]
+      }),
+      { total: 0, failed: 0, skipped: 0, success: 0 }
+    );
+    const rate = this.toRate(stats.success, stats.total - stats.skipped).toFixed(2);
+
+    return `Test cases: ${stats.total} (${rate}%, success: ${stats.success}, failed: ${stats.failed}, skipped: ${stats.skipped})`;
+  }
+
+  getTableData(results: ResultsType): Array<Array<string | number>> {
+    const header = ['No.', 'Path', 'Test cases', 'Successes', 'Failures', 'Skipped', 'Success rate'];
+    const rows = results.map((item, index) => {
+      const [path, total, success, failed, skipped, rate] = item;
+      return [index + 1, path, total, success, failed, skipped, `${rate.toFixed(2)}%`];
+    });
+
+    return [header, ...rows];
+  }
+
+  async onEnd(result: FullResult) {
+    const results = this.getResults();
+
+    console.log(this.formatTable(results));
+    console.log(this.getTestedPagesResult(results));
+    console.log(this.getTestCasesResult(results));
+    console.log(`Finished the run: ${result.status}`);
+
+    await this.saveResultsToFile(results, result.status);
+  }
+
+  formatTable(results: ResultsType): string {
+    const tableData = this.getTableData(results);
+    if (tableData.length === 0) {
+      return '';
+    }
+
+    const colWidths = tableData[0].map((_, colIndex) => {
+      let maxWidth = 0;
+      for (const row of tableData) {
+        const width = String(row[colIndex]).length;
+        if (width > maxWidth) {
+          maxWidth = width;
+        }
+      }
+      return maxWidth;
+    });
+
+    const lines: string[] = [];
+
+    const header = tableData[0].map((cell, i) => String(cell).padEnd(colWidths[i])).join(' │ ');
+    lines.push(header);
+
+    const separator = colWidths.map(width => '─'.repeat(width)).join('─┼─');
+    lines.push(separator);
+
+    for (let i = 1; i < tableData.length; i++) {
+      const row = tableData[i].map((cell, j) => String(cell).padEnd(colWidths[j])).join(' │ ');
+      lines.push(row);
+    }
+
+    return lines.join('\n');
+  }
+
+  async saveResultsToFile(results: ResultsType, status: string) {
+    const contents = [
+      this.formatTable(results),
+      this.getTestedPagesResult(results),
+      this.getTestCasesResult(results),
+      `Finished the run: ${status}`
+    ].join('\n');
+
+    try {
+      await fs.mkdir(cfg.outputPath, { recursive: true });
+      await fs.writeFile(`${cfg.outputPath}/${OUTPUT_FILE_NAME}`, contents, 'utf8');
```

Comment:
We could use `path.join` to combine the path parts, which could improve portability (e.g., on Windows).